### PR TITLE
Adds the tag-release tool

### PR DESCRIPTION
**Pull request by tcsc**
Adds a small tool to allow easy tagging of the root repository and every submodule in it at once. This will

a) help when different parts of the release process are trying to stitch
   together a Teleport release from the various submodules, and
b) provide useful historical data about which submodule code was included
   when, without people having to trace forward from the root repository

Example Usage: 

To tag the `teleport` repo and the top-level submodules, and push the tag to each repository's `origin` remote
```
> go run ./cmd/tag-release -repo ~/work/teleport -tag v123.4.5  -note "Optional annotation" -push
```

To tag the `teleport` repo and _all_ submodules (and not push anything anywhere):
```
> go run ./cmd/tag-release -recursive -repo ~/work/teleport -tag v123.4.5  -note "Optional annotation"
```

### CODE REVIEW

File: build.assets/tooling/cmd/tag-release/main.go [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+		}
+	}
+
+	for _, repo := range repos {
+		w, err := repo.Worktree()
+		if err != nil {
+			if !force {
+				return trace.Wrap(err, "Failed extracting worktree")
+			}
+			log.Printf("WARNING: failed extracting woirktree")
```
Comment (GavinFrazar) [inline build.assets/tooling/cmd/tag-release/main.go:122]: nit
```suggestion
			log.Printf("WARNING: failed extracting worktree")
```

File: build.assets/tooling/cmd/tag-release/main.go [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+		if t.Name().String() == tagpath {
+			res = true
+			return nil
+		}
+		return nil
+	})
+
+	if err != nil {
+		return false, trace.Wrap(err, "Failed enumerating tags")
+	}
```
Comment (GavinFrazar) [inline build.assets/tooling/cmd/tag-release/main.go:224]: Need to assign err above
```suggestion
	err = simpleTags.ForEach(func(t *plumbing.Reference) error {
		if t.Name().String() == tagpath {
			res = true
			return nil
		}
		return nil
	})

	if err != nil {
		return false, trace.Wrap(err, "Failed enumerating tags")
	}
```

File: build.assets/tooling/cmd/tag-release/main.go [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+		}
+
+		if tagExists {
+			if !force {
+				return trace.BadParameter("tag already exists")
+			}
+			log.Printf("WARNING: tag \"%s\" already exists. Deleting.", tag)
+			err = repo.DeleteTag(tag)
+			if err != nil {
+				log.Printf("WARNING: Faled to delete tag: %s", err)
```
Comment (GavinFrazar) [inline build.assets/tooling/cmd/tag-release/main.go:140]: nit
```suggestion
				log.Printf("WARNING: Failed to delete tag: %s", err)
```

File: build.assets/tooling/cmd/tag-release/main.go [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+			log.Printf("WARNING: failed extracting woirktree")
+			continue
+		}
+
+		log.Printf("Tagging %s", w.Filesystem.Root())
+
+		tagExists, err := statTag(repo, tag)
+		if err != nil {
+			return trace.Wrap(err, "failed enumerating tags")
+		}
```
Comment (GavinFrazar) [inline build.assets/tooling/cmd/tag-release/main.go:131]: nit: no need to add the extra message to the error, `statTag` already adds this message.
```suggestion
		tagExists, err := statTag(repo, tag)
		if err != nil {
			return trace.Wrap(err)
		}
```

File: build.assets/tooling/cmd/tag-release/main.go [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+
+		if tagExists {
+			if !force {
+				return trace.BadParameter("tag already exists")
+			}
+			log.Printf("WARNING: tag \"%s\" already exists. Deleting.", tag)
```
Comment (GavinFrazar) [inline build.assets/tooling/cmd/tag-release/main.go:137]: `%q` quotes the string for you
```suggestion
			log.Printf("WARNING: tag %q already exists. Deleting.", tag)
```

File: build.assets/tooling/cmd/tag-release/main.go [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+		head, err := repo.Head()
+		if err != nil {
+			return trace.Wrap(err, "failed fetching head revision")
+		}
+
+		log.Printf("Tagging %s => %s", w.Filesystem.Root(), head.String())
+		_, err = repo.CreateTag(tag, head.Hash(), opts)
+		if err != nil {
+			if !force {
+				return trace.Wrap(err, "")
```
Comment (GavinFrazar) [inline build.assets/tooling/cmd/tag-release/main.go:154]: nit
```suggestion
				return trace.Wrap(err)
```

**Review by GavinFrazar — COMMENTED** [review]
comments and suggestions

File: build.assets/tooling/cmd/tag-release/main.go (hunk 2 above)
Comment (GavinFrazar) [reply build.assets/tooling/cmd/tag-release/main.go:224]: although by the looks of it, error will always be `nil`? Might be able to just delete the error check if that's the case.

Comment (tcsc) [reply build.assets/tooling/cmd/tag-release/main.go:224]: I expect that the underlying enumeration can fail, even if our callbacks don't

**Review by GavinFrazar — APPROVED** [review]

**Comment by tcsc** [comment]
Overkill for the task required

**Pull request closed without merging**